### PR TITLE
fix #3722. Maintain global transform fixed when setting parent.

### LINKED ISSUE #3722: ofNode setParent maintaining global position doesn't work as expected
Both :
```
ofNode::setParent( p, true );
ofNode::setParent( p, false );
```
do the same. When `bMaintainGlobalTransform == true` we should compute the future transform the child will have and then set it. Right [now](https://github.com/openframeworks/openFrameworks/blob/31d24eac22f817534c2a9c17d6fe3b6d01957a70/libs/openFrameworks/3d/ofNode.cpp#L20) the transform is reevaluated from the new parent, hence modifying the global transform. 

Here's what I think it should be:

```
if(bMaintainGlobalTransform) {
    ofMatrix4x4 postParentGlobalTransform = parent.getGlobalTransformMatrix().getInverse() * getGlobalTransformMatrix();
    this->parent = &parent;
    setTransformMatrix(postParentGlobalTransform);
} else {
    this->parent = &parent;
}
```

[PR](https://github.com/openframeworks/openFrameworks/pull/3723) sent!

### PR DESCRIPTION
**Pull request by chuckleplant**
Replaces PR #3723

### CODE REVIEW

File: libs/openFrameworks/3d/ofNode.cpp [hunk 1]
```diff
@@ -15,12 +15,12 @@ ofNode::ofNode()
 //----------------------------------------
 void ofNode::setParent(ofNode& parent, bool bMaintainGlobalTransform) {
     if(bMaintainGlobalTransform) {
-        ofMatrix4x4 globalTransform(getGlobalTransformMatrix());
-        this->parent = &parent;
-        setTransformMatrix(globalTransform);
-    } else {
-        this->parent = &parent;
-    }
+		ofMatrix4x4 postParentGlobalTransform = getGlobalTransformMatrix() * parent.getGlobalTransformMatrix().getInverse();
```
Comment (arturoc) [inline libs/openFrameworks/3d/ofNode.cpp]: isn't this just the same as the else branch? multiplying the current global transformation by the inverse of the parent's global will just return the current local no?

Comment (chuckleplant) [inline libs/openFrameworks/3d/ofNode.cpp]: No because the incoming node is not yet the parent.

Comment (arturoc) [inline libs/openFrameworks/3d/ofNode.cpp]: oh, yeah you are right.

**Comment by arturoc** [comment]
looks correct to me but not sure why the original implementation was like that. ping @memo who i think did it.

**Comment by memo** [comment]
what seems to be the problem?

The idea was if you give a node a new parent, but tell it to maintain global transform, it needs to update the nodes *localtransform* according to its *new* parent transform to stay in the new global transform. Whereas if you tell it to NOT maintain global transform, it maintains its *local* transform, gets attached to the new parent, and hence will have a new global transform.

**Comment by memo** [comment]
oh i see, the issue is in setTransformMatrix(globalTransform);
weird, I'm sure I'd tested this when I wrote it.

I think the problem is in the ambiguity of setTransformMatrix. Is it setting local transform or global? (looking at the code I can see it is setting the local).

For everything else we have setGlobalXXX vs setLocalXXX. (And we have getLocalTransform and getGlobalTransform).

perhaps we should do the same for setLocalTransformMatrix and setGlobalTransformMatrix (and keep setTransformMatrix calling setLocalTransformMatrix for backwards compatibility, but deprecate it?)

**Comment by chuckleplant** [comment]
+1, there is indeed an ambiguity with that function.

**Comment by arturoc** [comment]
i'm going to merge this then and we can discuss if we need setGlobalTransformMatrix in a different issue. @chuckleplant or @memo can you open it?

**Comment by memo** [comment]
sure will do

**Pull request merged**